File: backend/app/services/request_service.py
from datetime import datetime
from decimal import Decimal
from typing import Optional, List
from sqlalchemy.orm import Session
from fastapi import HTTPException, status

from app.models.request import Request, RequestType, RequestStatus
from app.models.position import Position, PositionStatus
from app.schemas.request import BuyRequestCreate, SellRequestCreate, RequestApprove
from app.schemas.position import PositionCreate, PriceTarget
from app.services.position_service import PositionService


class RequestService:
    def __init__(self, db: Session):
        self.db = db
        self.position_service = PositionService(db)

    def get_request_by_id(self, request_id: int) -> Optional[Request]:
        return self.db.query(Request).filter(Request.id == request_id).first()

    def get_requests(
        self,
        status: Optional[str] = None,
        request_type: Optional[str] = None,
        requester_id: Optional[int] = None,
        page: int = 1,
        limit: int = 20
    ) -> tuple[List[Request], int]:
        query = self.db.query(Request)

        if status:
            query = query.filter(Request.status == status)
        if request_type:
            query = query.filter(Request.request_type == request_type)
        if requester_id:
            query = query.filter(Request.requester_id == requester_id)

        total = query.count()
        requests = query.order_by(Request.created_at.desc()).offset((page - 1) * limit).limit(limit).all()

        return requests, total

    def _convert_targets_to_json(self, targets) -> list:
        """Decimal을 float로 변환하여 JSON 직렬화 가능하게 함"""
        if not targets:
            return None
        result = []
        for t in targets:
            item = t.model_dump() if hasattr(t, 'model_dump') else t
            # Decimal을 float로 변환
            result.append({
                k: float(v) if isinstance(v, Decimal) else v
                for k, v in item.items()
            })
        return result

    def create_buy_request(self, request_data: BuyRequestCreate, requester_id: int) -> Request:
        request = Request(
            requester_id=requester_id,
            request_type=RequestType.BUY.value,
            target_ticker=request_data.target_ticker,
            ticker_name=request_data.ticker_name,
            target_market=request_data.target_market,
            order_type=request_data.order_type,
            order_amount=request_data.order_amount,
            order_quantity=request_data.order_quantity,
            buy_price=request_data.buy_price,
            buy_orders=None,  # Legacy field - not used
            target_ratio=request_data.target_ratio,
            take_profit_targets=self._convert_targets_to_json(request_data.take_profit_targets),
            stop_loss_targets=self._convert_targets_to_json(request_data.stop_loss_targets),
            memo=request_data.memo,
            status=RequestStatus.PENDING.value
        )

        self.db.add(request)
        self.db.commit()
        self.db.refresh(request)

        return request

    def create_sell_request(self, request_data: SellRequestCreate, requester_id: int) -> Request:
        # Verify position exists and is open
        position = self.position_service.get_position_by_id(request_data.position_id)
        if not position:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Position not found"
            )
        if position.status != PositionStatus.OPEN.value:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Position is not open"
            )

        request = Request(
            requester_id=requester_id,
            position_id=request_data.position_id,
            request_type=RequestType.SELL.value,
            target_ticker=position.ticker,
            target_market=position.market,
            sell_quantity=request_data.sell_quantity,
            sell_price=request_data.sell_price,
            sell_reason=request_data.sell_reason,
            status=RequestStatus.PENDING.value
        )

        self.db.add(request)
        self.db.commit()
        self.db.refresh(request)

        return request

    def _add_completed_flag_to_targets(self, targets: list, all_completed: bool = False) -> list:
        """타겟 리스트에 completed 플래그 추가"""
        if not targets:
            return None
        return [
            {**t, "completed": all_completed}
            for t in targets
        ]

    def _create_buy_plan(self, request) -> list:
        """요청에서 매수 계획 생성 (첫 번째만 완료 처리)"""
        buy_plan = []

        # 분할 매수 계획이 있는 경우
        if request.buy_orders and len(request.buy_orders) > 0:
            for i, order in enumerate(request.buy_orders):
                buy_plan.append({
                    "price": float(order.get("price", 0)),
                    "ratio": float(order.get("ratio", 0)),
                    "completed": (i == 0)  # 첫 번째만 완료
                })
        # 단일 매수인 경우
        elif request.buy_price and request.order_quantity:
            buy_plan.append({
                "price": float(request.buy_price),
                "quantity": float(request.order_quantity),
                "completed": True  # 단일 매수는 완료 처리
            })

        return buy_plan if buy_plan else None

    def approve_request(self, request_id: int, approve_data: RequestApprove, approved_by: int) -> tuple[Request, Optional[Position]]:
        request = self.get_request_by_id(request_id)
        if not request:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Request not found"
            )

        if request.status != RequestStatus.PENDING.value and request.status != RequestStatus.DISCUSSION.value:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Request cannot be approved (current status: {request.status})"
            )

        # 요청자의 희망가/수량 사용 (팀장이 별도 입력 안 하면)
        executed_price = approve_data.executed_price or request.buy_price
        executed_quantity = approve_data.executed_quantity or request.order_quantity

        # 매수 요청인데 가격/수량이 없으면 에러
        if request.request_type == RequestType.BUY.value:
            if not executed_price or not executed_quantity:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="매수 가격과 수량이 필요합니다"
                )

        request.status = RequestStatus.APPROVED.value
        request.approved_by = approved_by
        request.approved_at = datetime.utcnow()
        request.executed_price = executed_price
        request.executed_quantity = executed_quantity
        request.executed_at = approve_data.executed_at or datetime.utcnow()

        position = None

        if request.request_type == RequestType.BUY.value:
            # Check if open position exists for this ticker
            existing_position = self.position_service.get_open_position_by_ticker(
                request.target_ticker, request.target_market
            )

            # 익절/손절 타겟에 completed 플래그 추가 (모두 미완료)
            tp_targets = self._add_completed_flag_to_targets(request.take_profit_targets, False)
            sl_targets = self._add_completed_flag_to_targets(request.stop_loss_targets, False)

            # 매수 계획 생성
            buy_plan = self._create_buy_plan(request)

            if existing_position:
                # Add to existing position (기존 포지션에 추가 매수)
                position = self.position_service.add_to_position(
                    existing_position,
                    executed_quantity,
                    executed_price,
                    tp_targets,
                    sl_targets,
                    buy_plan
                )
            else:
                # Create new position (is_info_confirmed = False)
                position = self.position_service.create_position_from_request(
                    ticker=request.target_ticker,
                    ticker_name=request.ticker_name,
                    market=request.target_market,
                    buy_price=executed_price,
                    quantity=executed_quantity,
                    buy_plan=buy_plan,
                    take_profit_targets=tp_targets,
                    stop_loss_targets=sl_targets,
                    opened_by=request.requester_id
                )

            request.position_id = position.id

        elif request.request_type == RequestType.SELL.value:
            position = self.position_service.get_position_by_id(request.position_id)
            sell_price = approve_data.executed_price or request.sell_price
            sell_quantity = approve_data.executed_quantity or request.sell_quantity

            if not sell_quantity:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="매도 수량이 필요합니다"
                )

            if position:
                # Reduce position quantity
                remaining_quantity = (position.total_quantity or Decimal(0)) - sell_quantity

                if remaining_quantity <= 0:
                    # Close position completely
                    from app.schemas.position import PositionClose
                    close_data = PositionClose(
                        average_sell_price=sell_price,
                        total_sell_amount=sell_price * sell_quantity if sell_price else sell_quantity * (position.average_buy_price or Decimal(0)),
                        closed_at=approve_data.executed_at
                    )
                    position = self.position_service.close_position(
                        position.id, close_data, approved_by
                    )
                else:
                    # Reduce position
                    position = self.position_service.reduce_position(
                        position, sell_quantity
                    )

        self.db.commit()
        self.db.refresh(request)

        return request, position

    def reject_request(self, request_id: int, rejection_reason: str, rejected_by: int) -> Request:
        request = self.get_request_by_id(request_id)
        if not request:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Request not found"
            )

        if request.status != RequestStatus.PENDING.value and request.status != RequestStatus.DISCUSSION.value:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Request cannot be rejected (current status: {request.status})"
            )

        request.status = RequestStatus.REJECTED.value
        request.approved_by = rejected_by
        request.approved_at = datetime.utcnow()
        request.rejection_reason = rejection_reason

        self.db.commit()
        self.db.refresh(request)

        return request

    def start_discussion(self, request_id: int) -> Request:
        request = self.get_request_by_id(request_id)
        if not request:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Request not found"
            )

        if request.status != RequestStatus.PENDING.value:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Discussion cannot be started (current status: {request.status})"
            )

        request.status = RequestStatus.DISCUSSION.value

        self.db.commit()
        self.db.refresh(request)

        return request

File: spring-backend/backend/src/main/java/com/fundmessenger/request/service/RequestService.java
package com.fundmessenger.request.service;

import com.fundmessenger.common.exception.BusinessException;
import com.fundmessenger.common.exception.NotFoundException;
import com.fundmessenger.common.util.TargetConverter;
import com.fundmessenger.position.dto.PositionClose;
import com.fundmessenger.position.entity.Position;
import com.fundmessenger.position.service.PositionService;
import com.fundmessenger.request.dto.BuyRequestCreate;
import com.fundmessenger.request.dto.RequestApprove;
import com.fundmessenger.request.dto.SellRequestCreate;
import com.fundmessenger.request.entity.TradeRequest;
import com.fundmessenger.request.repository.TradeRequestRepository;
import com.fundmessenger.user.entity.User;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@Slf4j
@Service
@RequiredArgsConstructor
public class RequestService {

    private final TradeRequestRepository tradeRequestRepository;
    private final PositionService positionService;

    // ──────────────────────────────────────────────
    // Read operations
    // ──────────────────────────────────────────────

    /**
     * Get a single request by ID.
     */
    public TradeRequest getRequestById(Long requestId) {
        return tradeRequestRepository.findById(requestId)
                .orElseThrow(() -> new NotFoundException("Request", requestId));
    }

    /**
     * Get paginated and filtered request list.
     */
    public Page<TradeRequest> getRequests(
            String status,
            String requestType,
            Long requesterId,
            int page,
            int limit
    ) {
        PageRequest pageable = PageRequest.of(page - 1, limit, Sort.by(Sort.Direction.DESC, "createdAt"));

        // Use specification-style filtering via repository
        if (status != null && requestType != null && requesterId != null) {
            return tradeRequestRepository.findByStatusAndRequestTypeAndRequesterId(
                    status, requestType, requesterId, pageable);
        } else if (status != null && requestType != null) {
            return tradeRequestRepository.findByStatusAndRequestType(status, requestType, pageable);
        } else if (status != null && requesterId != null) {
            return tradeRequestRepository.findByStatusAndRequesterId(status, requesterId, pageable);
        } else if (requestType != null && requesterId != null) {
            return tradeRequestRepository.findByRequestTypeAndRequesterId(requestType, requesterId, pageable);
        } else if (status != null) {
            return tradeRequestRepository.findByStatus(status, pageable);
        } else if (requestType != null) {
            return tradeRequestRepository.findByRequestType(requestType, pageable);
        } else if (requesterId != null) {
            return tradeRequestRepository.findByRequesterId(requesterId, pageable);
        } else {
            return tradeRequestRepository.findAll(pageable);
        }
    }

    // ──────────────────────────────────────────────
    // Create operations
    // ──────────────────────────────────────────────

    /**
     * Create a buy request.
     */
    @Transactional
    public TradeRequest createBuyRequest(BuyRequestCreate dto, User requester) {
        TradeRequest request = new TradeRequest();
        request.setRequester(requester);
        request.setRequestType("buy");
        request.setTargetTicker(dto.getTargetTicker());
        request.setTickerName(dto.getTickerName());
        request.setTargetMarket(dto.getTargetMarket());
        request.setOrderType(dto.getOrderType());
        request.setOrderAmount(dto.getOrderAmount());
        request.setOrderQuantity(dto.getOrderQuantity());
        request.setBuyPrice(dto.getBuyPrice());
        request.setBuyOrders(null); // Legacy field - not used
        request.setTargetRatio(dto.getTargetRatio());
        request.setTakeProfitTargets(TargetConverter.convertTargets(dto.getTakeProfitTargets()));
        request.setStopLossTargets(TargetConverter.convertTargets(dto.getStopLossTargets()));
        request.setMemo(dto.getMemo());
        request.setStatus("pending");

        TradeRequest saved = tradeRequestRepository.save(request);
        log.info("Buy request created: id={}, ticker={}, by={}",
                saved.getId(), dto.getTargetTicker(), requester.getUsername());
        return saved;
    }

    /**
     * Create a sell request. Validates that the position exists and is open.
     */
    @Transactional
    public TradeRequest createSellRequest(SellRequestCreate dto, User requester) {
        Position position = positionService.getPositionById(dto.getPositionId());

        if (!"open".equals(position.getStatus())) {
            throw new BusinessException("Position is not open");
        }

        TradeRequest request = new TradeRequest();
        request.setRequester(requester);
        request.setPosition(position);
        request.setRequestType("sell");
        request.setTargetTicker(position.getTicker());
        request.setTargetMarket(position.getMarket());
        request.setSellQuantity(dto.getSellQuantity());
        request.setSellPrice(dto.getSellPrice());
        request.setSellReason(dto.getSellReason());
        request.setStatus("pending");

        TradeRequest saved = tradeRequestRepository.save(request);
        log.info("Sell request created: id={}, ticker={}, qty={}, by={}",
                saved.getId(), position.getTicker(), dto.getSellQuantity(), requester.getUsername());
        return saved;
    }

    // ──────────────────────────────────────────────
    // Approval / Rejection / Discussion
    // ──────────────────────────────────────────────

    /**
     * Approve a request. For buy requests, creates or adds to a position.
     * For sell requests, reduces or closes a position.
     *
     * @return a two-element array: [TradeRequest, Position (may be null for edge cases)]
     */
    @Transactional
    public Object[] approveRequest(Long requestId, RequestApprove approveData, User approver) {
        TradeRequest request = getRequestById(requestId);

        if (!"pending".equals(request.getStatus()) && !"discussion".equals(request.getStatus())) {
            throw new BusinessException(
                    "Request cannot be approved (current status: " + request.getStatus() + ")");
        }

        // Fall back to requester's desired price/quantity if manager didn't provide overrides
        BigDecimal executedPrice = approveData.getExecutedPrice() != null
                ? approveData.getExecutedPrice() : request.getBuyPrice();
        BigDecimal executedQuantity = approveData.getExecutedQuantity() != null
                ? approveData.getExecutedQuantity() : request.getOrderQuantity();

        // For buy requests, price and quantity are mandatory
        if ("buy".equals(request.getRequestType())) {
            if (executedPrice == null || executedQuantity == null) {
                throw new BusinessException("Buy price and quantity are required (매수 가격과 수량이 필요합니다)");
            }
        }

        // Update request status
        request.setStatus("approved");
        request.setApprover(approver);
        request.setApprovedAt(OffsetDateTime.now());
        request.setExecutedPrice(executedPrice);
        request.setExecutedQuantity(executedQuantity);
        request.setExecutedAt(approveData.getExecutedAt() != null
                ? approveData.getExecutedAt() : OffsetDateTime.now());

        Position position = null;

        if ("buy".equals(request.getRequestType())) {
            position = handleBuyApproval(request, executedPrice, executedQuantity);
        } else if ("sell".equals(request.getRequestType())) {
            position = handleSellApproval(request, approveData, approver);
        }

        TradeRequest saved = tradeRequestRepository.save(request);
        return new Object[]{saved, position};
    }

    /**
     * Reject a request.
     */
    @Transactional
    public TradeRequest rejectRequest(Long requestId, String rejectionReason, User rejector) {
        TradeRequest request = getRequestById(requestId);

        if (!"pending".equals(request.getStatus()) && !"discussion".equals(request.getStatus())) {
            throw new BusinessException(
                    "Request cannot be rejected (current status: " + request.getStatus() + ")");
        }

        request.setStatus("rejected");
        request.setApprover(rejector);
        request.setApprovedAt(OffsetDateTime.now());
        request.setRejectionReason(rejectionReason);

        TradeRequest saved = tradeRequestRepository.save(request);
        log.info("Request {} rejected by {}, reason: {}",
                requestId, rejector.getUsername(), rejectionReason);
        return saved;
    }

    /**
     * Update request status to "discussion".
     */
    @Transactional
    public TradeRequest startDiscussion(Long requestId) {
        TradeRequest request = getRequestById(requestId);

        if (!"pending".equals(request.getStatus())) {
            throw new BusinessException(
                    "Discussion cannot be started (current status: " + request.getStatus() + ")");
        }

        request.setStatus("discussion");
        TradeRequest saved = tradeRequestRepository.save(request);
        log.info("Request {} status changed to discussion", requestId);
        return saved;
    }

    // ──────────────────────────────────────────────
    // Private helpers
    // ──────────────────────────────────────────────

    @SuppressWarnings("unchecked")
    private Position handleBuyApproval(
            TradeRequest request,
            BigDecimal executedPrice,
            BigDecimal executedQuantity
    ) {
        // Check for existing open position
        Position existingPosition = positionService.getOpenPositionByTicker(
                request.getTargetTicker(), request.getTargetMarket()).orElse(null);

        // Add completed flag to targets (all uncompleted)
        List<Map<String, Object>> tpTargets = addCompletedFlagToTargets(
                (List<Map<String, Object>>) request.getTakeProfitTargets(), false);
        List<Map<String, Object>> slTargets = addCompletedFlagToTargets(
                (List<Map<String, Object>>) request.getStopLossTargets(), false);

        // Create buy plan from request
        List<Map<String, Object>> buyPlan = createBuyPlan(request);

        Position position;
        if (existingPosition != null) {
            // Add to existing position
            position = positionService.addToPosition(
                    existingPosition, executedQuantity, executedPrice,
                    tpTargets, slTargets, buyPlan);
        } else {
            // Create new position
            position = positionService.createPositionFromRequest(
                    request.getTargetTicker(),
                    request.getTickerName(),
                    request.getTargetMarket(),
                    executedPrice,
                    executedQuantity,
                    buyPlan,
                    tpTargets,
                    slTargets,
                    request.getRequester().getId());
        }

        request.setPosition(position);
        return position;
    }

    private Position handleSellApproval(
            TradeRequest request,
            RequestApprove approveData,
            User approver
    ) {
        Position position = request.getPosition() != null
                ? positionService.getPositionById(request.getPosition().getId())
                : null;

        BigDecimal sellPrice = approveData.getExecutedPrice() != null
                ? approveData.getExecutedPrice() : request.getSellPrice();
        BigDecimal sellQuantity = approveData.getExecutedQuantity() != null
                ? approveData.getExecutedQuantity() : request.getSellQuantity();

        if (sellQuantity == null) {
            throw new BusinessException("Sell quantity is required (매도 수량이 필요합니다)");
        }

        if (position != null) {
            BigDecimal remainingQuantity = (position.getTotalQuantity() != null
                    ? position.getTotalQuantity() : BigDecimal.ZERO).subtract(sellQuantity);

            if (remainingQuantity.compareTo(BigDecimal.ZERO) <= 0) {
                // Close position completely
                BigDecimal totalSellAmount = sellPrice != null
                        ? sellPrice.multiply(sellQuantity)
                        : sellQuantity.multiply(
                        position.getAverageBuyPrice() != null
                                ? position.getAverageBuyPrice() : BigDecimal.ZERO);

                PositionClose closeData = new PositionClose();
                closeData.setTotalSellAmount(totalSellAmount);
                closeData.setAverageSellPrice(sellPrice);
                closeData.setClosedAt(approveData.getExecutedAt());
                position = positionService.closePosition(
                        position.getId(), closeData, approver);
            } else {
                // Reduce position
                position = positionService.reducePosition(position, sellQuantity);
            }
        }

        return position;
    }

    /**
     * Adds a "completed" flag to each target entry.
     */
    @SuppressWarnings("unchecked")
    private List<Map<String, Object>> addCompletedFlagToTargets(
            List<Map<String, Object>> targets,
            boolean allCompleted
    ) {
        if (targets == null || targets.isEmpty()) {
            return null;
        }
        List<Map<String, Object>> result = new ArrayList<>();
        for (Map<String, Object> t : targets) {
            Map<String, Object> copy = new LinkedHashMap<>(t);
            copy.put("completed", allCompleted);
            result.add(copy);
        }
        return result;
    }

    /**
     * Creates a buy plan from the request's buy orders or single buy info.
     * First entry is marked as completed.
     */
    @SuppressWarnings("unchecked")
    private List<Map<String, Object>> createBuyPlan(TradeRequest request) {
        List<Map<String, Object>> buyPlan = new ArrayList<>();

        // Split buy orders (분할 매수 계획)
        Object rawBuyOrders = request.getBuyOrders();
        if (rawBuyOrders instanceof List && !((List<?>) rawBuyOrders).isEmpty()) {
            List<Map<String, Object>> buyOrders = (List<Map<String, Object>>) rawBuyOrders;
            for (int i = 0; i < buyOrders.size(); i++) {
                Map<String, Object> order = buyOrders.get(i);
                Map<String, Object> entry = new LinkedHashMap<>();
                entry.put("price", toDouble(order.get("price")));
                entry.put("ratio", toDouble(order.get("ratio")));
                entry.put("completed", i == 0); // First one completed
                buyPlan.add(entry);
            }
        }
        // Single buy
        else if (request.getBuyPrice() != null && request.getOrderQuantity() != null) {
            Map<String, Object> entry = new LinkedHashMap<>();
            entry.put("price", request.getBuyPrice().doubleValue());
            entry.put("quantity", request.getOrderQuantity().doubleValue());
            entry.put("completed", true); // Single buy is completed
            buyPlan.add(entry);
        }

        return buyPlan.isEmpty() ? null : buyPlan;
    }

    private double toDouble(Object value) {
        if (value == null) return 0.0;
        if (value instanceof Number) return ((Number) value).doubleValue();
        try {
            return Double.parseDouble(value.toString());
        } catch (NumberFormatException e) {
            return 0.0;
        }
    }
}
